### ISSUE TITLE
Support N instances of enrichments in a graph

### ISSUE THREAD

**Issue by chuwy** [snowplow/snowplow#2260]
Currently we have two types of enrichments:
- built-in enrichments: currency, ip lookup, weather, etc
- custom enrichments: JS enrichment, API lookup enrichment (WIP)

Custom enrichment imply it need to be flexible and user can have several instances of each (eg. 3 API lookups + 2 JS).

**Comment by chuwy** [comment]
One of quick possible solutions is to use arrays in JS and API lookups configuration JSONs.

**Comment by alexanderdean** [comment]
This ticket also needs to factor in the dependency graph between instances. We have a couple of options here:

* Crude dependency tree: *EnrichmentC depends on EnrichmentB and EnrichmentA*
* Precise dependency tree: *EnrichmentC depends on EnrichmentB.outputs.user.ip_address and EnrichmentA.outputs.derived_tstamp*

**Comment by claudiomartins** [comment]
+1 on this 

This is incredibly important for multi-layered enhancers - we are starting to work with multiple types of events (struct/unstruct) that would be very useful to created dependencies among them and multiple enhancers of the same type (mostly API).

**Comment by alexanderdean** [comment]
As a quick update - I am leaning strongly towards the precise dependency tree here. I think it's achievable and much preferable.

**Comment by thomasintveld** [comment]
The ability to use multiple enrichments of the same type in parallel is very important — and one could argue that SQL Enrichment is not fully supported without it. I.e. you now have to choose whether you enrich events with user-data, or with data about the page they are viewing. +1 on this

**Comment by szareiangm** [comment]
This ticket is a large amount of work that in nature they are separated: 

1) The fist part that @chuwy mentioned as an example, is a far more simpler than the graph dependency of enrichments. However, it needs another nesting around enrichments to give them a name. This is actually very helpful, too. We can have  user defined names wrapping around enrichments to reuse them. 
To avoid the issues with parsing Json config files, we can have the enrichment tasks names inside them as an optional field and convert the configuration to an array.
```
[{
"userDefinedName" : "my_company_api_lookup"   // usual API lookup configuration
// the rest of the config for this API
} , {
userDefinedName: "acme_api_lookup"    // another usual API lookup configuration
// the rest of the config for this API
}]
```
2) For the second part, we need to tear apart the enrichment process and add a flexible/dynamic 
task (enrichment) scheduler that reads its plan from a DAG file. This part can use the previous one, with calling tasks with the "user-defined" names instead of their vanilla name. To me, this is a huge overhaul. 

I think for the sake of getting things done, we should at least split this ticket into two. What do you guys think?

**Comment by BenFradet** [comment]
I think adding names to enrichments as a separate ticket and leaving this one for a real dependency-driven dag is fine :+1:

**Comment by chuwy** [comment]
Migrated to https://github.com/snowplow/enrich/issues/130
